File: app/src/main/java/io/treehouses/remote/Fragments/SSHConsole.java
package io.treehouses.remote.Fragments;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.text.ClipboardManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;

import com.google.android.material.tabs.TabLayout;

import io.treehouses.remote.Constants;
import io.treehouses.remote.R;
import io.treehouses.remote.SSH.PromptHelper;
import io.treehouses.remote.SSH.Terminal.TerminalBridge;
import io.treehouses.remote.SSH.Terminal.TerminalManager;
import io.treehouses.remote.SSH.Terminal.TerminalView;
import io.treehouses.remote.SSH.Terminal.TerminalViewPager;
import io.treehouses.remote.SSH.interfaces.BridgeDisconnectedListener;
import io.treehouses.remote.bases.BaseFragment;
import io.treehouses.remote.bases.BaseSSHConsole;
import io.treehouses.remote.databinding.ActivitySshConsoleBinding;

public class SSHConsole extends BaseSSHConsole implements BridgeDisconnectedListener {
    private ActivitySshConsoleBinding bind;

    private static final int KEYBOARD_DISPLAY_TIME = 3000;
    private static final int KEYBOARD_REPEAT_INITIAL = 500;
    private static final int KEYBOARD_REPEAT = 100;

    @Nullable
    protected TerminalManager bound = null;
//    protected TerminalPagerAdapter adapter = null;
    protected LayoutInflater inflater = null;

    private SharedPreferences prefs = null;

    // determines whether or not menuitem accelerators are bound
    // otherwise they collide with an external keyboard's CTRL-char
    private boolean hardKeyboard = false;

    protected Uri requested;

    protected ClipboardManager clipboard;

    private LinearLayout keyboardGroup;
    private Runnable keyboardGroupHider;

    private TextView empty;

    private Animation fade_out_delayed;

    private Animation keyboard_fade_in, keyboard_fade_out;

    private MenuItem disconnect, copy, paste, resize, urlscan;

    private boolean forcedOrientation;

    private Handler handler = new Handler();

    private View contentView;

    private ImageView mKeyboardButton;

    private ServiceConnection connection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName className, IBinder service) {
            bound = ((TerminalManager.TerminalBinder) service).getService();

            // let manager know about our event handling services
            bound.disconnectListener = SSHConsole.this;
            bound.setResizeAllowed(true);

            final String requestedNickname = (requested != null) ? requested.getFragment() : null;
            TerminalBridge requestedBridge = bound.getConnectedBridge(requestedNickname);

            // If we didn't find the requested connection, try opening it
            if (requestedNickname != null && requestedBridge == null) {
                try {
                    Log.d("TAG", String.format("We couldnt find an existing bridge with URI=%s (nickname=%s), so creating one now", requested.toString(), requestedNickname));
                    requestedBridge = bound.openConnection(requested);
                } catch (Exception e) {
                    Log.e("TAG", "Problem while trying to create new requested bridge from URI", e);
                }
            }

            // create views for all bridges on this service
            adapter.notifyDataSetChanged();
            final int requestedIndex = bound.getBridges().indexOf(requestedBridge);

            if (requestedBridge != null)
                requestedBridge.promptHelper.setHandler(promptHandler);


            if (requestedIndex != -1) {
                pager.post(() -> setDisplayedTerminal(requestedIndex));
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName className) {
            bound = null;
            adapter.notifyDataSetChanged();
            updateEmptyVisible();
        }
    };


    private void setDisplayedTerminal(int requestedIndex) {
        pager.setCurrentItem(requestedIndex);
        // set activity title
        onTerminalChanged();
    }

    private void onTerminalChanged() {
        View terminalNameOverlay = findCurrentView(R.id.terminal_name_overlay);
        if (terminalNameOverlay != null)
            terminalNameOverlay.startAnimation(fade_out_delayed);
        updateDefault();
        updatePromptVisible();
        ActivityCompat.invalidateOptionsMenu(requireActivity());
    }
    protected View findCurrentView(int id) {
        View view = pager.findViewWithTag(adapter.getBridgeAtPosition(pager.getCurrentItem()));
        if (view == null) {
            return null;
        }
        return view.findViewById(id);
    }

    private void updateDefault() {
        // update the current default terminal
        TerminalView view = adapter.getCurrentTerminalView();
        if (view == null || bound == null) {
            return;
        }
        bound.defaultBridge = view.bridge;
    }

    protected void updateEmptyVisible() {
        // update visibility of empty status message
        empty.setVisibility((pager.getChildCount() == 0) ? View.VISIBLE : View.GONE);
    }


    @Override
    public void onStart() {
        super.onStart();

        // connect with manager service to find all bridges
        // when connected it will insert all views
        requireActivity().bindService(new Intent(getActivity(), TerminalManager.class), connection, Context.BIND_AUTO_CREATE);
    }

    @Override
    public void onStop() {
        super.onStop();

        requireActivity().unbindService(connection);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        bind = ActivitySshConsoleBinding.inflate(inflater, container, false);
        return bind.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        onLoad(mHandler);
    }

    @SuppressLint("HandlerLeak")
    private final Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if (msg.what == Constants.MESSAGE_READ) {
                String readMessage = (String) msg.obj;
                Log.d("TAG", "readMessage = " + readMessage);
//                performAction(readMessage);
            }
        }
    };

    @Override
    public void onDisconnected(TerminalBridge bridge) {

    }
}

File: app/src/main/kotlin/io/treehouses/remote/sshconsole/SSHConsole.kt
/*
 * ConnectBot: simple, powerful, open-source SSH client for Android
 * Copyright 2007 Kenny Root, Jeffrey Sharkey
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.treehouses.remote.sshconsole

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.ActivityInfo
import android.content.res.Configuration
import android.media.AudioManager
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.os.StrictMode
import android.text.ClipboardManager
import android.view.*
import android.widget.HorizontalScrollView
import androidx.core.view.MenuItemCompat
import androidx.preference.PreferenceManager
import io.treehouses.remote.InitialActivity
import io.treehouses.remote.PreferenceConstants
import io.treehouses.remote.R
import io.treehouses.remote.ssh.terminal.TerminalBridge
import io.treehouses.remote.ssh.terminal.TerminalManager
import io.treehouses.remote.ssh.terminal.TerminalManager.TerminalBinder
import io.treehouses.remote.ssh.interfaces.BridgeDisconnectedListener
import io.treehouses.remote.databinding.ActivitySshConsoleBinding
import io.treehouses.remote.utils.logD
import io.treehouses.remote.utils.logE

open class SSHConsole : DerivedSSHConsole(), BridgeDisconnectedListener {

    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            bound = (service as TerminalBinder).service

            // let manager know about our event handling services
            bound!!.disconnectListener = this@SSHConsole
            bound!!.isResizeAllowed = true
            val requestedNickname = if (requested != null) requested!!.fragment else null
            var requestedBridge = bound!!.mNicknameBridgeMap[requestedNickname]?.get()

            // If we didn't find the requested connection, try opening it
            if (requestedNickname != null && requestedBridge == null) {
                try {
                    logD("${String.format("We couldnt find an existing bridge with URI=%s (nickname=%s), so creating one now", requested.toString(), requestedNickname)}")
                    requestedBridge = bound!!.openConnection(requested)
                } catch (e: Exception) {
                    logE("Problem while trying to create new requested bridge from URI $e")
                }
            }

            // create views for all bridges on this service
            adapter!!.notifyDataSetChanged()
            val requestedIndex = bound!!.bridges.indexOf(requestedBridge)
            requestedBridge?.promptHelper?.setHandler(promptHandler)
            if (requestedIndex != -1) {
                bind.pager.post { setDisplayedTerminal(requestedIndex) }
            }
        }

        override fun onServiceDisconnected(className: ComponentName) {
            bound = null
            adapter!!.notifyDataSetChanged()
            updateEmptyVisible()
        }
    }

    override fun onDisconnected(bridge: TerminalBridge) {
        synchronized(adapter!!) {
            adapter!!.notifyDataSetChanged()
            logD("Someone sending HANDLE_DISCONNECT to parentHandler")
            if (bridge.isAwaitingClose) {
                closeBridge()
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    public override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)
        StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.LAX)
        hardKeyboard = resources.configuration.keyboard == Configuration.KEYBOARD_QWERTY
        clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        titleBarHide = prefs!!.getBoolean(PreferenceConstants.TITLEBARHIDE, false)
        if (titleBarHide) requestActionBar()
        bind = ActivitySshConsoleBinding.inflate(layoutInflater)
        setContentView(bind.root)

        // handle requested console from incoming intent
        if (icicle == null) { requested = intent.data
        } else {
            val uri = icicle.getString(STATE_SELECTED_URI)
            if (uri != null) requested = Uri.parse(uri)
        }
        setUpTerminalPager()
        empty = findViewById(android.R.id.empty)
        promptListeners()
        setUpKeyboard()
        addKeyboardListeners()
        keyboardScroll()
        setUpActionBar()

        detectSoftVisibility()

    }

    @SuppressLint("ClickableViewAccessibility")
    private fun keyboardScroll() {
        val keyboardScroll = findViewById<HorizontalScrollView>(R.id.keyboard_hscroll)
        if (!hardKeyboard) {
            // Show virtual keyboard and scroll back and forth
            showEmulatedKeys(false)
            keyboardScroll.postDelayed({
                val xscroll = bind.keyboard.buttonF12.right
                keyboardScroll.smoothScrollBy(xscroll, 0)
                keyboardScroll.postDelayed({ keyboardScroll.smoothScrollBy(-xscroll, 0) }, 500)
            }, 500)
        }

        // Reset keyboard auto-hide timer when scrolling
        keyboardScroll.setOnTouchListener { v: View, event: MotionEvent ->
            when (event.action) {
                MotionEvent.ACTION_MOVE -> autoHideEmulatedKeys()
                MotionEvent.ACTION_UP -> {
                    v.performClick()
                    return@setOnTouchListener true
                }
            }
            false
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        val view = adapter!!.currentTerminalView
        val activeTerminal = view != null
        var (sessionOpen, disconnected) = checkSession(view, activeTerminal)
        menu.setQwertyMode(true)
        disconnect = menu.add("Disconnect")
        if (hardKeyboard) disconnect!!.alphabeticShortcut = 'w'
        if (!sessionOpen && disconnected) disconnect!!.title = "Close Console"
        disconnect!!.isEnabled = activeTerminal
        disconnect!!.setIcon(android.R.drawable.ic_menu_close_clear_cancel)
        setDisconnectItemListener()
        paste = menu.add("Paste")
        if (hardKeyboard) paste!!.alphabeticShortcut = 'v'
        MenuItemCompat.setShowAsAction(paste, MenuItemCompat.SHOW_AS_ACTION_IF_ROOM)
        paste!!.isEnabled = activeTerminal
        setPasteItemListener()
        urlScan = menu.add("Scan for URLs")
        if (hardKeyboard) urlScan!!.alphabeticShortcut = 'u'
        urlScan!!.setIcon(android.R.drawable.ic_menu_search)
        urlScan!!.isEnabled = activeTerminal
        setUrlItemListener()
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        super.onPrepareOptionsMenu(menu)
        volumeControlStream = AudioManager.STREAM_NOTIFICATION
        val view = adapter?.currentTerminalView
        val activeTerminal = view != null
        var (sessionOpen, disconnected) = checkSession(view, activeTerminal)
        disconnect?.isEnabled = activeTerminal
        if (sessionOpen || !disconnected) disconnect?.title = "Disconnect" else disconnect?.title = "Close Console"
        paste?.isEnabled = activeTerminal
        //		portForward.setEnabled(sessionOpen && canForwardPorts);
        urlScan?.isEnabled = activeTerminal
        resize?.isEnabled = sessionOpen
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item == null) return false
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    public override fun onStart() {
        super.onStart()

        // connect with manager service to find all bridges
        // when connected it will insert all views
        bindService(Intent(this, TerminalManager::class.java), connection, Context.BIND_AUTO_CREATE)
    }

    public override fun onPause() {
        super.onPause()
        logD("onPause called")
        if (forcedOrientation && bound != null) {
            bound!!.isResizeAllowed = false
        }
    }

    public override fun onResume() {
        super.onResume()
        logD("onResume called")

        // Make sure we don't let the screen fall asleep.
        // This also keeps the Wi-Fi chipset from disconnecting us.
        if (prefs!!.getBoolean(PreferenceConstants.KEEP_ALIVE, true)) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
        configureOrientation()
        if (forcedOrientation && bound != null) {
            bound!!.isResizeAllowed = true
        }
    }

    /* (non-Javadoc)
     * @see android.app.Activity#onNewIntent(android.content.Intent)
     */
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        requested = intent.data
        if (requested == null) {
            logD("Got null intent data in onNewIntent()")
            return
        }
        if (bound == null) {
            logD("We're not bound in onNewIntent()")
            return
        }
        setIntentRequested()
    }

    public override fun onStop() {
        super.onStop()
        unbindService(connection)
    }

    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
        // Maintain selected host if connected.
        val currentTerminalView = adapter!!.currentTerminalView
        if (currentTerminalView != null
                && !currentTerminalView.bridge.isDisconnected) {
            requested = currentTerminalView.bridge.host!!.uri
            savedInstanceState.putString(STATE_SELECTED_URI, requested.toString())
        }
        super.onSaveInstanceState(savedInstanceState)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        logD("${String.format("onConfigurationChanged; requestedOrientation=%d, newConfig.orientation=%d", requestedOrientation, newConfig.orientation)}")
        if (bound != null) {
            bound!!.isResizeAllowed = !(forcedOrientation &&
                    (newConfig.orientation != Configuration.ORIENTATION_LANDSCAPE &&
                            requestedOrientation == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE ||
                            newConfig.orientation != Configuration.ORIENTATION_PORTRAIT &&
                            requestedOrientation == ActivityInfo.SCREEN_ORIENTATION_PORTRAIT))
            bound!!.hardKeyboardHidden = newConfig.hardKeyboardHidden == Configuration.HARDKEYBOARDHIDDEN_YES
            bind.keyboard.buttonKeyboard.visibility = if (bound!!.hardKeyboardHidden) View.VISIBLE else View.GONE
        }
    }
}
